### DISCUSSION TITLE
Pkl package in a private gitlab repo

### DISCUSSION THREAD

**Discussion by KyleReczek-AbsorbLMS** [apple/pkl#632] [Q&A] [answered]
What is the recommended way of using a package that is hosted in a private repo using gitlab/github/etc? 

I'm having some issues because it requires auth. My project sync gets the following error. 

'Received unexpected status code `403` when making request `GET https://gitlab.com/users/sign_in`.'

Is there a standard way people are doing this currently?

**Comment by bioball** [comment, marked as answer]
At the moment, there is no way to use authentication to fetch a package. This is something that makes sense for Pkl to support, though. Feel free to file a GitHub issue for it!

**Reply by KyleReczek-AbsorbLMS** [reply]
Darn! I guess I'll use my workaround of git submodules for now. Thanks!

**Reply by KyleReczek-AbsorbLMS** [reply]
Raised issue -> https://github.com/apple/pkl/issues/633

**Comment by shurshun** [comment]
I would suggest you to try this wrapper: https://github.com/hpklio/hpkl. It was developed for internal needs and is currently in the alpha stage, but it already allows you to store packages in OCI, as well as supports Vals as a custom reader

**Reply by bioball** [reply]
Interesting! What problems are you looking to solve with hpkl?

**Reply by shurshun** [reply]
As I mentioned, the primary problems we’re looking to solve with hpkl include the ability to store pkl packages in an OCI registry and support for a custom reader from the [helmfile/vals](https://github.com/helmfile/vals) project.
